src/versionize/src/version_map.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Maps struct/enum/union versions to a sequence of root versions.
+//! This is required to support the concept of a snapshot version
+//! composed of individually versioned components.
+
+use std::any::TypeId;
+use std::collections::hash_map::HashMap;
+
+const BASE_VERSION: u16 = 1;
+
+///
+/// The VersionMap API provides functionality to define the version for each serialized
+/// type and attach them to specific root versions.
+///
+/// !TODO: Find an O(1) solution for `get_type_version()`.
+///
+#[derive(Clone, Debug, Default)]
+pub struct VersionMap {
+    versions: Vec<HashMap<TypeId, u16>>,
+}
+
+impl VersionMap {
+    /// Create a new version map and set root version to 1.
+    pub fn new() -> Self {
+        VersionMap {
+            versions: vec![HashMap::new(); 1],
+        }
+    }
+
+    /// Bumps root version by 1 to create a new root version and set it as latest version.
+    pub fn new_version(&mut self) -> &mut Self {
+        self.versions.push(HashMap::new());
+        self
+    }
+
+    /// Define a mapping between a specific type version and the latest root version.
+    pub fn set_type_version(&mut self, type_id: TypeId, type_version: u16) -> &mut Self {
+        // It is safe to unwrap since `self.versions` always has at least 1 element.
+        self.versions
+            .last_mut()
+            .unwrap()
+            .insert(type_id, type_version);
+        self
+    }
+
+    /// Returns the version of `type_id` corresponding to the specified `root_version`.
+    pub fn get_type_version(&self, root_version: u16, type_id: TypeId) -> u16 {
+        let version_space = self.versions.split_at(root_version as usize).0;
+
+        for i in (0..version_space.len()).rev() {
+            if let Some(version) = version_space[i].get(&type_id) {
+                return *version;
+            }
+        }
+
+        BASE_VERSION
+    }
+
+    /// Returns the latest top version.
+    pub fn latest_version(&self) -> u16 {
+        self.versions.len() as u16
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
This should be `mod tests {`.

Reply:
I will update this for consistency.

Reply:
It needs to be tests because otherwise kcov is not ignoring the section and you end up with bogus coverage results. You can see how the tests section is excluded from the coverage report here: https://github.com/firecracker-microvm/firecracker/blob/e6436c599756b653931a57d5f5c418f7d22cac40/tests/integration_tests/build/test_coverage.py#L61

Reply:
A mechanism to enforce this might also be useful. I just did a search and there are some other place where this happens:
src/cpuid/src/transformer/intel.rs
src/cpuid/src/transformer/mod.rs
src/cpuid/src/transformer/amd.rs

Reply:
Yap, a mechanism would be nice indeed. Can you add an issue about this? And one for fixing the current problems?

Reply:
Sure.

Reply:
fixed.